[Gloo] GPU alltoall

jiayisuse opened the pull request:
Summary: Gloo alltoall for GPU

Test Plan: buck test mode/dev-nosan caffe2/torch/lib/c10d:ProcessGroupGlooTest

Differential Revision: D22631554

rohan-varma commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2305,8 +2305,78 @@ class AsyncAlltoallWork : public ProcessGroupGloo::AsyncWork {
+    outputEvents.front().block(at::cuda::getCurrentCUDAStream());
+  }
+
+  at::Tensor tmpOutput;
```
nit: would these be better named `cpuInput`/`cpuOutput` to reflect that they are CPU tensors?

rohan-varma commented on torch/lib/c10d/test/ProcessGroupGlooTest.cpp:
```diff
@@ -555,4 +555,13 @@ TEST(ProcessGroupGlooTest, testBroadcastCUDA) {
+  {
+    if (torch::cuda::is_available()) {
+      TemporaryFile file;
+      testBroadcast(file.path, at::DeviceType::CUDA);
```
Am I missing something or should this test the `allToall` functionality?

jiayisuse replied:
oops published from a different dev box

jiayisuse committed:
GLOO process group GPU alltoall (#41690)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/41690

Gloo alltoall for GPU

Test Plan: buck test mode/dev-nosan caffe2/torch/lib/c10d:ProcessGroupGlooTest

Reviewed By: osalpekar

Differential Revision: D22631554

fbshipit-source-id: 22a4e132741cdd9e7ac69f7c816e1339ae47dce9
